### DISCUSSION TITLE
Storing internal state / non-mapped attributes on mapped ORM classes

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10923] [Usage Questions]
After scouring through the documentation, I'm having trouble figuring out what the recommended approach is for storing internal state (non-mapped attributes) on every mapped class in my ORM, which all inherit from DeclarativeBase.

Obviously with a regular Python class this would be done with a `self.attribute` data member that is initialized in the `__init__`, but when subclassing DeclarativeBase, special things are happening with the `__init__` that make it difficult to determine what the best practice is here.

Some things I've tried:

1. Implement custom DeclarativeBase constructor:
```python
class Base(DeclarativeBase):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.internal_state = {}
```
This doesn't work due to the special behavior of `__init__` and `super()` when dealing with DeclarativeBase.

2. Use a mixin class:
```python
class Base(DeclarativeBase):
    pass

class Mixin:
    def __init__(self):
        self.internal_state = {}

class MyMappedClass(Base, Mixin):
    id: Mapped[int] = mapped_column(primary_key=True)
```
Here the `Mixin.__init__` seems to never be called when creating a new instance of `MyMappedClass`

3. Convert to dataclass
```python
class Base(MappedAsDataclass, DeclarativeBase):
    internal_state: dict = field(default_factory=dict)
```
I was able to get this approach to work but I really don't want to convert all of my ORM classes to dataclasses if I can avoid it, because it makes everything more cumbersome. I have to re-order the declaration of all of my Mapped columns since columns without defaults must come before ones with defaults. And worse, every column that was formerly declared like this:
```python
attribute: Mapped[str | None]
```
when using dataclasses has to be become:
```python
attribute: Mapped[str | None] = mapped_column(default=None)
```

The key criteria here is that I *don't* want to disrupt the normal behavior of `__init__` that you get with DeclarativeBase -- I have no need to augment the args/kwargs passed in on construction, I simply need a place to initialize some internal data members whenever an ORM instance is created.

**Comment by zzzeek** [comment]
it's a little unfortunate the `__init__` on the `Base` doesn't work, we had some issues with this and I thought that was fixed.

in any case, mixin is close, use an abstract:

```py
class Base(DeclarativeBase):
    pass

class InitBase(Base):
    __abstract__ = True

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.internal_state = {}

class A(InitBase):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
```

alternatively use the [init event](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.init)

```py
class Base(DeclarativeBase):
    pass

@event.listens_for(Base, "init", propagate=True)
def receive_init(target, args, kwargs):
    target.internal_state = {}

class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

```

**Reply by dkbarn** [reply]
So the problem (I think) with the `__abstract__` approach that you've given above is that `self.internal_state` will only be bound to Python-local instances of `A`. If you query for `A` objects and try to access `self.internal_state` on them, you'll get an AttributeError.

Does the "init event" approach address that issue? Edit: Just read the docs for init event.. it does not 😔

It seems like I will need to define both an init event and load event to handle both userland-constructed instances and db-fetched instances.

**Reply by zzzeek** [reply]
well we were talking about init, there's a whole different story for when the objects load from the database.  you'd want to add in the use of the [load](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.load) and possibly the [refresh](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.InstanceEvents.refresh) events also.

we are lacking docs on this use right now which is a mistake, they got removed and were not restored.  I will restore them now.   I agree with @CaselIT below it's much better to use a cached_property.

**Reply by foobacca** [reply]
For anyone else who ends up here looking for an answer, if you want to use the event approach when you are using mapping imperatively, you can do:

``` python
metadata = MetaData()

products = Table(
    "products",
    metadata,
    Column("id", Integer, primary_key=True),
    Column("data", String(255)),
)

mapper_reg = registry()
mapper_reg.map_imperatively(model.Product, products)

@event.listens_for(model.Product, "load")
def receive_load(product, _):
    product.internal_state = {}
```

And in the `model` module you can just have:

``` python
class Product:
    def __init__(self, id: int, data: str):
        self.id = id
        self.data = data
        self.internal_state = {}
```

Then you can write states against the `Product` class directly without worrying about SQLAlchemy. But when you run integration tests and the class is created by SQLAlchemy with data from the database, you will still get `internal_state` set up properly.

**Comment by CaselIT** [comment]
Hi,

If you only need to initialize some objects there you may also look into the stdlib cached_property to create the attribute on first access

**Reply by dkbarn** [reply]
This turned out to be the simplest approach to get working, thanks for the suggestion!
